core/embed/trezorhal/unix/flash.c

```diff
@@ -28,12 +28,81 @@
 
 #include "common.h"
 #include "flash.h"
+#include "model.h"
 #include "profile.h"
 
 #ifndef FLASH_FILE
 #define FLASH_FILE profile_flash_path()
 #endif
 
+#if defined STM32F427xx || defined STM32F429xx
+#define FLASH_SECTOR_COUNT 24
+#elif defined STM32F405x
+#define FLASH_SECTOR_COUNT 12
```

Comment:
I think this is meant for model 1 which uses STM32F405x headers even though it's actually F2. Probably best to change the ifdef to use TREZOR_MODEL.

Reply:
hm do you see benefit in keeping the unix implementation of flash driver mimicking STM32F4/F2? this will get even more messier when we add U5 models as the flash is quite different and with the abstraction implemented we might not care anymore about keeping the sectors 1:1 with the MCU used. thinking about removing the ifdefs all together and just keeping single implementation/layout for unix.

Reply:
Firstly, just so we are on the same page, my comment was basically saying that `FLASH_SECTOR_COUNT` for `STM32F405x` is incompatible with the definitions of `STORAGE_AREAS`, `FIRMWARE_AREA` etc. in this file. So these need to be defined differently for `STM32F405x`. I would try to avoid redefining them and instead reuse the definitions from `core/embed/models/model_T1T1_layout.c` and `core/embed/models/model_T2T1_layout.c` so that the layout in the hardware build is always the same as that in the unix build and we don't have to update it in two places.

As for the higher-level question of whether we should keep the sectors 1:1 with the MCU used, I don't know. In principle we should do our best to keep the emulator as close to the hardware implementation as possible, so that we detect bugs as early as possible. On the other hand I am not sure whether the sector layout is significant in that regard. I can think of two things that are relevant:
- The differences in storage size for different models. I believe we should keep the storage sector sizes 1:1 with hardware, so that our device tests detect if some new feature starts to clutter the storage too much.
- The firmware hashing functionality that reads the firmware sectors. That would have to be somehow hacked in the emulator, so it wouldn't get tested very well by the device tests.

Finally the question of keeping the unix implementation of the flash driver that mimicks STM32F4/F2. Are we ever going to run `core` firmware on T1? I guess it's nice to keep, since we already invested some resources into this effort, but not sure if it has any future that would be worth the maintenance. Nevertheless, the problems that we are discovering here are going to come up sooner or later with newer MCU models, so we might as well try and solve them early, while the code is fresh in our minds.

Reply:
reusing the layout files: https://github.com/trezor/trezor-firmware/commit/557f0b871716334520a19b8a64898794663ddd64

this most likely wont be possible to keep this simple when adding U5 support. if we want to keep this close to real hw implementation there will be some larger ifdefs.

i don't think that T1 in core is relevant anymore, i would be ok with dropping it. it doesn't work anyway for several reasons. but that is for @matejcik to decide and out of scope here.